javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -0,0 +1,844 @@
+                ranked = Ranked.of(root1, rank(tree1) + 1);
+                children = concat(tree2, tree1.getChildren());
+            } else {
+                ranked = Ranked.of(root2, rank(tree2) + 1);
+                children = concat(tree1, tree2.getChildren());
+            }
+            return Tree.of(ranked, children);
+        }
+
+        /**
+         * fun skewLink (t0 as Node (x0,r0, _), t1 as Node (x1,r1,c1), t2 as Node (x2,r2,c2)) =
+         * *  if Elem.leq (x1,x0) andalso Elem.leq (x1,x2) then      Node (x1,r1+1,t0 :: t2 :: c1)
+         * *  else if Elem.leq (x2,x0) andalso Elem.leq (x2,x1) then Node (x2,r2+1,t0 :: t1 :: c2)
+         * *  else                                                   Node (x0,r1+1,[t1, t2])
+         **/
+        private static <T> Node<Ranked<T>> skewLink(SerializableComparator<? super T> comparator, Node<Ranked<T>> tree0, Node<Ranked<T>> tree1, Node<Ranked<T>> tree2) {
+            final T root0 = root(tree0), root1 = root(tree1), root2 = root(tree2);
+
+            final Ranked<T> ranked;
+            final List<Node<Ranked<T>>> children;
+            if (comparator.isLessOrEqual(root1, root0) && comparator.isLessOrEqual(root1, root2)) {
+                ranked = Ranked.of(root1, rank(tree1) + 1);
+                children = concat(tree0, concat(tree2, tree1.getChildren()));
+            } else if (comparator.isLessOrEqual(root2, root0) && comparator.isLessOrEqual(root2, root1)) {
+                ranked = Ranked.of(root2, rank(tree2) + 1);
+                children = concat(tree0, concat(tree1, tree2.getChildren()));
+            } else {
+                ranked = Ranked.of(root0, rank(tree1) + 1);
+                children = List.of(tree1, tree2);
+            }
+            return Tree.of(ranked, children);
+        }
+
+        /**
+         * fun findMin [] = raise EMPTY
+         * * | findMin [t] = root t
+         * * | findMin (t :: ts) =
+         * *     let val x = findMin ts
+         * *     in if Elem.leq (root t, x) then root t else x end
+         */
+        protected static <T> T findMin(SerializableComparator<? super T> comparator, List<Node<Ranked<T>>> forest, T currentMin) {
+            while (true) {
+                if (forest.size() == 1) {
+                    return currentMin;
+                } else {
+                    forest = forest.tail();
+                    currentMin = min(comparator, currentMin, root(forest.head()));
+                }
+            }
+        }
+
+        private static <T> T min(SerializableComparator<? super T> comparator, T value1, T value2) {
+            return comparator.isLessOrEqual(value1, value2) ? value1 : value2;
+        }
+
+        protected static <T> T root(Node<Ranked<T>> tree) {
+            return tree.getValue().value;
+        }
+
+        private static <T> int rank(Node<Ranked<T>> tree) {
+            return tree.getValue().rank;
+        }
+
+        private static <T> List<T> concat(T head, List<T> rest) {
```

Comment:
Please check its usages, it makes the code a lot more readable (and will be JITed anyway).
e.g.
```java
concat(tree0, concat(tree1, tree2.getChildren()));
```
instead of
```java
tree2.getChildren().prepend(tree1).prepend(tree0);
```
where the order is all wrong and there are no whitespaces delimiting the separate parts.